### ISSUE TITLE
Feature request: bind heap to arena such that arena increases memory when full

### ISSUE THREAD

**Issue by kuffnr01** [microsoft/mimalloc#784]
Dear mimalloc team, I saw the below response regarding #617 -- I very much like that feature to bind heaps to an arena to only allocate memory from the arena

But in my application I would need to be able to increase the size of an arena either explicitly or automatically (increase arena if full instead of running out of memory) -- would that be possible and if yes how? Thank you!!

Previous response 617
The latest dev-slice (v2) has an initial experimental api where you can create a heap that exclusively allocates in a certain arena (running out of memory when the arena is full). 

// Experimental: heaps associated with specific memory arena's
typedef int mi_arena_id_t;
mi_decl_export void* mi_arena_area(mi_arena_id_t arena_id, size_t* size);
mi_decl_export int   mi_reserve_huge_os_pages_at_ex(size_t pages, int numa_node, size_t timeout_msecs, bool exclusive, mi_arena_id_t* arena_id) mi_attr_noexcept;
mi_decl_export int   mi_reserve_os_memory_ex(size_t size, bool commit, bool allow_large, bool exclusive, mi_arena_id_t* arena_id) mi_attr_noexcept;
mi_decl_export bool  mi_manage_os_memory_ex(void* start, size_t size, bool is_committed, bool is_large, bool is_zero, int numa_node, bool exclusive, mi_arena_id_t* arena_id) mi_attr_noexcept;

#if MI_MALLOC_VERSION >= 200
mi_decl_nodiscard mi_decl_export mi_heap_t* mi_heap_new_in_arena(mi_arena_id_t arena_id);
#endif

**Comment by biddisco** [comment]
This issue is marked as closed, but there is no information on how to make a custom arena grow-able.
Is there further information on what to do to allow a custom arena to increase in size rather than throw an out of memory error when it runs out of space?

**Comment by FoxieFlakey** [comment]
Replied on a message by @biddisco
> ...what to do to allow a custom arena to increase in size rather than throw an out of memory error when it runs out of space?

TLDR, Add `mi_heap_add_arena` and `mi_request_memory_fun` so custom arena can insert more arena into heap when mimalloc requested.

First thing is let custom arena know if mimalloc needs more memory. My idea on that is a per-heap `mi_request_memory_fun` will be called when mimalloc request more memory (e.g. out of memory, somehow too fragmented for whatever reason, allocation failure, etc).

1. Try allocate normally on heap
2. If success, return the result
3. Call the `mi_request_memory_fun` function
4. If it return false, return allocation failure
5. Try allocate normally on heap
6. Return the result regardless if it failed or success

Prototype of `mi_request_memory_fun` function would be
```c
// Return true, if function would like to let
// mimalloc retries the allocation
typedef bool (*mi_request_memory_fun)(mi_heap_t* heapWhereItHasFailed, size_t requestedSize, void* userdata);
```

Second is the how to expand memory. On this, my idea is to have new function `mi_heap_add_arena` which adds new arena to existing heap. The prototype would be
```c
int mi_heap_add_arena(mi_heap_t* heap, mi_arena_id_t newArena);
```
